pkg/controller/cassandradatacenter/controller/statefulset.go

```diff
@@ -101,8 +97,12 @@ func CreateOrUpdateStatefulSet(
 
 				log.Info("No Cassandra nodes have been decommissioned. Decommissioning the newest one " + newestPod.Name)
 
-				if _, err := sidecar.NewSidecarClient(newestPod.Status.PodIP, &sidecarClientOptions).DecommissionNode(); err != nil {
-					return errors.New("Failure occurred while decommissioning a Cassandra node upon scaling a cluster down: " + err.Error())
+				if clientForNewestPod := sidecar.ClientForPod(clients, newestPod); clientForNewestPod != nil {
```

Comment:
`clients` is a map, with a key type of `corev1.Pod*`. `newestPod` is also a `corev1.Pod*`.
AFAIK, pointers are comparable, so you should just be able to do `clients[newestPod]` to get the client for the newest pod.

Reply:
it doesnt work like that ive tried that approach, i can revisit though

Reply:
@zegelin 

`newestPod` here is not a pointer. `allPods` is just an array `[]corev1.Pod` (Kubernetes API returns that). Then we want to know the newest pod so I do `allPods[len(allPods)-1]`. This is not a pointer again. Hence I can not use it as a key to the map.

Now `SidecarClients` returns a map of pointers to pods as keys and pointers to clients as values while it is taking that array of pods from Kubernetes. Check this:

    func SidecarClients(pods []corev1.Pod, clientOptions ClientOptions) map[*corev1.Pod]*Client {

        podClients := make(map[*corev1.Pod]*Client)

        for _, pod := range pods {
            key := pod                      // <----- see this?
            podClients[&key] = NewSidecarClient(pod.Status.PodIP, &clientOptions)
        }

	return podClients
}

Now without `key := pod`, if I was about to put a client like `podClients[pod]`, it would, interestingly, overwrite what I put there, because Go uses same address for `pod` in that loop. Hence I was ending up with just one entry in that map after the loop was done.

Check this where it is explained in depth:

https://stackoverflow.com/questions/48826460/using-pointers-in-a-for-loop-golang